### PR TITLE
Marker Cluster Plugin

### PR DESCRIPTION
**Pull request by tobrun**
This PR ports 2 open-source clustering algorithms from [googlemaps/android-maps-utils](https://github.com/googlemaps/android-maps-utils/tree/master/library/src/com/google/maps/android/clustering).
 - Non-hierarchical distance based algorithm
 - Grid based algorithm

ezgif com-video-to-gif 52

cc @mapbox/android

### CODE REVIEW

File: plugins/cluster/build.gradle
Comment (zugaldia) [inline plugins/cluster/build.gradle]: Was this left commented on purpose? Should we remove it instead?

File: plugins/cluster/src/main/AndroidManifest.xml
Comment (zugaldia) [inline plugins/cluster/src/main/AndroidManifest.xml]: I believe we can remove the application tag from here.

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/NonHierarchicalDistanceBasedAlgorithm.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/NonHierarchicalDistanceBasedAlgorithm.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java]: Remove `MapboxMap.OnMarkerDragListener`.

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java]: Remove?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java]: Remove?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java]: Remove?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

File: plugins/cluster/src/test/java/com/mapbox/mapboxsdk/plugins/cluster/ExampleUnitTest.java
Comment (zugaldia) [inline plugins/cluster/src/test/java/com/mapbox/mapboxsdk/plugins/cluster/ExampleUnitTest.java]: Looks like we can remove this file.

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (zugaldia) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: Ticket?

**Review by zugaldia — COMMENTED** [review]
Just small notes. I realize there're a bunch of `TODO` and `UNSUPPORTED` that probably come from the original library. I'd check which ones are worth ticketing. Otherwise, let's at least add a note saying something like "This TODO item was part of the original library (http://...), we welcome contributions if you want to work on it. See https://github.com/mapbox/mapbox-plugins-android#contributing for details."

File: plugins/cluster/build.gradle
Comment (tobrun) [inline plugins/cluster/build.gradle]: This setup was copied from the traffic plugin and will be required for release. Since there is a need for a final 5.1.0 of the map SDK (dependency on new OnCameraIdle API). I would leave this untouched until the release of this plugin.

File: plugins/cluster/src/main/AndroidManifest.xml
Comment (tobrun) [inline plugins/cluster/src/main/AndroidManifest.xml]: yes 👍

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/MarkerManager.java]: https://github.com/mapbox/mapbox-gl-native/issues/2450

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/NonHierarchicalDistanceBasedAlgorithm.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/NonHierarchicalDistanceBasedAlgorithm.java]: This is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/algo/PreCachingAlgorithmDecorator.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

File: plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java
Comment (tobrun) [inline plugins/cluster/src/main/java/com/mapbox/mapboxsdk/plugins/cluster/clustering/view/DefaultClusterRenderer.java]: this is a todo from the original implementation

**Comment by tobrun** [comment]
@zugaldia I was able to make the code work with a configurable flag for MarkerView though the performance wasn't great. I'm thinking on punting on Markers as we aim to deprecate MarkerView in upcoming releases in https://github.com/mapbox/mapbox-gl-native/issues/9365.

**Comment by mdakram** [comment]
Small issue with this plugin is there is no option to disable clustering at certain zoom level, say we need to plot 5 markers in 5 mtrs radius even at last zoom level we are not able to see all markers.

Is there any workaround for this?

**Comment by zugaldia** [comment]
> I'm thinking on punting on Markers as we aim to deprecate MarkerView in upcoming releases in mapbox/mapbox-gl-native#9365.

👍 on reducing the scope to markers for now.

**Review by zugaldia — APPROVED** [review]

**Comment by simon-the-canadian** [comment]
Is there a timeframe on getting this PR merged? I'm currently in the process of migrating from Google Maps and this is one of the main features missing on MapBox for my usage.

**Comment by tobrun** [comment]
@simon-the-canadian I'm working on rebasing this PR on latest changes of master. 
Will merge once CI has approved.

cc @lilykaiser

**Comment by simon-the-canadian** [comment]
@tobrun Are you guys planning on releasing an artifact to maven soonish?

**Comment by maheshchakkarwar1** [comment]
When I run this code. Not working for me.
org.maplibre.android.exceptions.CalledFromWorkerThreadException: Map interactions should happen on the UI thread. Method invoked from wrong thread is latLngForPixel. at org.maplibre.android.maps.NativeMapView.checkState(NativeMapView.java:119) at org.maplibre.android.maps.NativeMapView.latLngForPixel(NativeMapView.java:713) at org.maplibre.android.maps.Projection.fromScreenLocation(Projection.java:96) at org.maplibre.android.maps.Projection.getVisibleRegion(Projection.java:153) at org.maplibre.android.maps.Projection.getVisibleRegion(Projection.java:122) at commichelin.clustering.view.DefaultClusterRenderer$RenderTask.run(DefaultCl

**Pull request merged**
